src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -272,9 +283,21 @@ std::shared_ptr<AbstractLQPNode> AbstractLQPNode::_shallow_copy(LQPNodeMapping&
   return shallow_copy;
 }
 
-void AbstractLQPNode::_remove_output_pointer(const std::shared_ptr<AbstractLQPNode>& output) {
-  const auto iter =
-      std::find_if(_outputs.begin(), _outputs.end(), [&](const auto& other) { return output == other.lock(); });
+void AbstractLQPNode::_remove_output_pointer(const AbstractLQPNode& output) {
+  const auto iter = std::find_if(_outputs.begin(), _outputs.end(), [&](const auto& other) {
+    /**
+     * HACK! We're checking for other.expired() here as well when looking for `output`
+     * If nothing else breaks the only way we might get `other.expired()` to be true is if `other` is the expired
+     * weak_ptr<> to output - and thus the element we're looking for - in the following scenario:
+     *
+     * auto node_a = Node::make()
+     * auto node_b = Node::make(..., node_a)
+     *
+     * node_b.reset(); // ~AbstractLQPNode() will call `node_a_remove_output_pointer(node_b)`
+     *                 // But we can't lock node_b anymore, since its ref count is already 0
+     */
+    return &output == other.lock().get() || other.expired();
```

Comment:
Does this make sense? If it is expired, `other.lock()` will return an empty `shared_ptr`, so isn't the expired check superfluous?

Reply:
Is what you are suggesting that this can be simplified to 

```c++

    const auto locked_other = other.lock();
    return &output == other.lock().get() || !locked_other;
```
???

I adjusted it accordingly

Reply:
Why not `return &output == other.lock().get()`?

Reply:
This... would not be true if `other` is expired? `&output` will be `!= nullptr`, even if this gets called from `output`'s  destructor?

Reply:
Aaaah, I misunderstood the original version. In that case, I think the original one is easier to read. Maybe add a comment "Check if `other` either points to output or to an already deleted node"?

Reply:
Reverted to `other.expired()` check. I think, however, that the 10 line comment above this check doesn't need further expansion